File: resultados.py
from typing import TYPE_CHECKING, List, Optional, Any
from dataclasses import dataclass, field
import matplotlib.pyplot as plt

if TYPE_CHECKING:
    from labirinto import Estado

    
@dataclass
class ResultadoBusca:
    algoritmo: str
    encontrado: bool
    caminho: List["Estado"] = field(default_factory=list)
    acoes: List[str] = field(default_factory=list)
    nos_explorados: int = 0
    nos_expandidos: int = 0
    estados_explorados: List["Estado"] = field(default_factory=list)
    tempo_execucao: Any = 0.0
    tamanho_fronteira: int = 0
    custo_total: float = 0.0

    @property
    def tamanho_caminho(self) -> Optional[int]:
        return len(self.acoes) if self.encontrado else None

    def imprimir_metricas(self):
        print("\n------------------------------------------------------\n")
        print(f'Resultados obtidos:\n')
        print(f'Algoritmo executado: {self.algoritmo}')
        print(f'Solução encontrada: {"sim" if self.encontrado else "não"}')
        print(f'Nós explorados: {self.nos_explorados}')
        print(f'Nós expandidos: {self.nos_expandidos}')
        print(f'Tamanho do caminho encontrado: {self.tamanho_caminho}')
        print(f'Custo total do caminho: {self.custo_total}')
        print(f'Tempo de execução: {self.tempo_execucao:.4f}')
        print(f'Tamanho máximo da fronteira: {self.tamanho_fronteira}')


@dataclass
class ResultadoSimulatedAnnealing(ResultadoBusca):

    historico: List[float] = field(default_factory=list)
    melhor_custo: float = 0.0
    pior_custo: float = 0.0
    custo_medio: float = 0.0
    tempo_medio: float = 0.0
    iteracoes_medias: float = 0.0
    temperatura_inicial: float = 0.0
    temperatura_final: float = 0.0
    fator_resfriamento: float = 0.0
    quantidade_execucoes: int = 0
    melhor_solucao: Optional[List["Estado"]] = field(default_factory=list)
    taxa_sucesso: float = 0.0
    taxa_melhora: float = 0.0

    def imprimir_metricas(self):
        print("\n------------------------------------------------------\n")
        print(f'Resultados obtidos:\n')
        print(f'Algoritmo executado: {self.algoritmo}')
        print(f'Solução encontrada: {"sim" if self.encontrado else "não"}')
        print(f'Número de execuções: {self.quantidade_execucoes}')
        print(f'Temperatura inicial: {self.temperatura_inicial}')
        print(f'Temperatura final: {self.temperatura_final}')
        print(f'Fator de resfriamento: {self.fator_resfriamento}')
        print(f'Melhor custo encontrado: {self.melhor_custo}')
        print(f'Pior custo encontrado: {self.pior_custo}')
        print(f'Custo médio: {self.custo_medio:.4f}')
        print(f'Tempo médio: {self.tempo_medio:.4f}')
        print(f'Iterações médias: {self.iteracoes_medias}')
        print(f'Taxa de sucesso: {self.taxa_sucesso * 100:.2f}%')
        print(f'Melhor solução encontrada: {self.melhor_solucao}')
        print(f'Taxa de melhora: {self.taxa_melhora * 100:.2f}%')

    
    def plotar_convergencia(self):
        
        plt.figure(figsize=(10,5))
        plt.plot(self.historico)
        plt.xlabel("Iteração")
        plt.ylabel("Custo")

        plt.title(
            "Simulated Annealing - Curva de Convergência"
        )

        plt.grid(True)

        plt.show()
        plt.savefig("grafico-sa.png")


@dataclass
class ResultadoHillClimbing(ResultadoBusca):
    melhor_custo: float = 0.0
    pior_custo: float = 0.0
    custo_medio: float = 0.0
    tempo_medio: float = 0.0
    iteracoes_medias: float = 0.0
    quantidade_execucoes: int = 0
    melhor_solucao: Optional[List["Estado"]] = field(default_factory=list)
    taxa_sucesso: float = 0.0
    taxa_melhora: float = 0.0

    def imprimir_metricas(self):
        print("\n------------------------------------------------------\n")
        print(f'Resultados obtidos por Múltiplas Execuções:\n')
        print(f'Algoritmo executado: {self.algoritmo}')
        print(f'Solução encontrada na melhor execução: {"sim" if self.encontrado else "não"}')
        print(f'Número de execuções: {self.quantidade_execucoes}')
        print(f'Melhor custo encontrado: {self.melhor_custo}')
        print(f'Pior custo encontrado: {self.pior_custo}')
        print(f'Custo médio: {self.custo_medio:.4f}')
        print(f'Tempo médio: {self.tempo_medio:.4f}')
        print(f'Iterações médias (Passos): {self.iteracoes_medias:.2f}')
        print(f'Taxa de sucesso (solução aceitável): {self.taxa_sucesso * 100:.2f}%')
        print(f'Taxa de melhora sobre o início: {self.taxa_melhora * 100:.2f}%')
        print(f'Melhor ordem de visitação encontrada: {self.melhor_solucao}')
@dataclass
class ResultadoBuscaOnline:
    algoritmo: str
    encontrado: bool
    caminho: List["Estado"] = field(default_factory=list)
    acoes: List[str] = field(default_factory=list)
    movimentos: int = 0
    custo_real: float = 0.0
    celulas_reveladas: int = 0
    celulas_revisitadas: int = 0
    replanejamentos: int = 0
    custo_otimo_offline: float = 0.0
    razao_online_offline: float = 0.0
    tempo_execucao: float = 0.0

    @property
    def tamanho_caminho(self) -> Optional[int]:
        return len(self.acoes) if self.encontrado else None

    def imprimir_metricas(self):
        print("\n------------------------------------------------------\n")
        print(f"Resultados obtidos:\n")
        print(f"Algoritmo executado: {self.algoritmo}")
        print(f"Solução encontrada: {'sim' if self.encontrado else 'não'}")
        print(f"Número total de movimentos: {self.movimentos}")
        print(f"Custo real percorrido: {self.custo_real}")
        print(f"Número de células reveladas: {self.celulas_reveladas}")
        print(f"Número de células revisitadas: {self.celulas_revisitadas}")
        print(f"Número de replanejamentos: {self.replanejamentos}")
        print(f"Tamanho do caminho encontrado: {self.tamanho_caminho}")
        print(f"Custo ótimo offline: {self.custo_otimo_offline}")
        print(f"Razão online/offline: {self.razao_online_offline:.4f}")
        print(f"Tempo de execução: {self.tempo_execucao:.4f}")

File: interface-web/scripts/resultados.js
export class ResultadoBusca {
  constructor({
    algoritmo = "",
    encontrado = false,
    caminho = [],
    acoes = [],
    nosExplorados = 0,
    nosExpandidos = 0,
    estadosExplorados = [],
    tempoExecucao = 0.0,
    tamanhoFronteira = 0,
    custoTotal = 0.0,
  } = {}) {
    this.algoritmo = algoritmo;
    this.encontrado = encontrado;
    this.caminho = caminho;
    this.acoes = acoes;
    this.nosExplorados = nosExplorados;
    this.nosExpandidos = nosExpandidos;
    this.estadosExplorados = estadosExplorados;
    this.tempoExecucao = tempoExecucao;
    this.tamanhoFronteira = tamanhoFronteira;
    this.custoTotal = custoTotal;
  }

  get tamanhoCaminho() {
    return this.encontrado ? this.acoes.length : null;
  }
}

export class ResultadoSimulatedAnnealing extends ResultadoBusca {
  constructor({
    historico = [],
    melhorCusto = 0.0,
    piorCusto = 0.0,
    custoMedio = 0.0,
    tempoMedio = 0.0,
    iteracoesMedias = 0.0,
    taxaSucesso = 0.0,
    temperaturaInicial = 0.0,
    temperaturaFinal = 0.0,
    fatorResfriamento = 0.0,
    quantidadeExecucoes = 0,
    melhorSolucao = [],
    taxaMelhora = 0.0,
    ...resultadoBusca
  } = {}) {
    super(resultadoBusca);
    this.historico = historico;
    this.melhorCusto = melhorCusto;
    this.piorCusto = piorCusto;
    this.custoMedio = custoMedio;
    this.tempoMedio = tempoMedio;
    this.iteracoesMedias = iteracoesMedias;
    this.taxaSucesso = taxaSucesso;
    this.temperaturaInicial = temperaturaInicial;
    this.temperaturaFinal = temperaturaFinal;
    this.fatorResfriamento = fatorResfriamento;
    this.quantidadeExecucoes = quantidadeExecucoes;
    this.melhorSolucao = melhorSolucao;
    this.taxaMelhora = taxaMelhora;
  }

  plotarConvergencia() {
    const canvas = document.getElementById("graficoConvergencia");

    const labels = this.historico.map((_, indice) => indice);

    if (window.graficoSA) {
      window.graficoSA.destroy();
    }

    window.graficoSA = new Chart(canvas, {
      type: "line",

      data: {
        labels: labels,

        datasets: [
          {
            label: "Custo",

            data: this.historico,

            borderColor: "rgb(54, 162, 235)",

            backgroundColor: "rgba(54, 162, 235, 0.2)",

            tension: 0.2,

            fill: false,
          },
        ],
      },

      options: {
        responsive: true,

        plugins: {
          title: {
            display: true,
            text: "Simulated Annealing - Curva de Convergência",
          },
        },

        scales: {
          x: {
            title: {
              display: true,
              text: "Iteração",
            },
          },

          y: {
            title: {
              display: true,
              text: "Custo",
            },
          },
        },
      },
    });
  }
}

export class ResultadoHillClimbing extends ResultadoBusca {
  constructor({
    historico = [],
    melhorCusto = 0.0,
    piorCusto = 0.0,
    custoMedio = 0.0,
    tempoMedio = 0.0,
    iteracoesMedias = 0.0,
    quantidadeExecucoes = 0,
    melhorSolucao = [],
    taxaSucesso = 0.0,
    taxaMelhora = 0.0,
    ...resultadoBusca
  } = {}) {
    super(resultadoBusca);
    this.historico = historico
    this.melhorCusto = melhorCusto;
    this.piorCusto = piorCusto;
    this.custoMedio = custoMedio;
    this.tempoMedio = tempoMedio;
    this.iteracoesMedias = iteracoesMedias;
    this.quantidadeExecucoes = quantidadeExecucoes;
    this.melhorSolucao = melhorSolucao;
    this.taxaSucesso = taxaSucesso;
    this.taxaMelhora = taxaMelhora;
  }
}

export class ResultadoBuscaOnline extends ResultadoBusca {
  constructor({
    algoritmo,
    encontrado,
    caminho = [],
    acoes = [],
    movimentos = 0,
    custoReal = 0.0,
    celulasReveladas = 0,
    celulasRevisitadas = 0,
    replanejamentos = 0,
    custoOtimoOffline = 0.0,
    razaoOnlineOffline = 0.0,
    tempoExecucao = 0.0,
    ...resultado
  }) {
    super(resultado);
    this.algoritmo = algoritmo;
    this.encontrado = encontrado;

    this.caminho = caminho;
    this.acoes = acoes;

    this.movimentos = movimentos;
    this.custoReal = custoReal;

    this.celulasReveladas = celulasReveladas;
    this.celulasRevisitadas = celulasRevisitadas;

    this.replanejamentos = replanejamentos;

    this.custoOtimoOffline = custoOtimoOffline;
    this.razaoOnlineOffline = razaoOnlineOffline;

    this.tempoExecucao = tempoExecucao;
  }

  get tamanhoCaminho() {
    return this.encontrado ? this.acoes.length : null;
  }
}
